modules/pytket-iqm/pytket/extensions/iqm/backends/iqm.py

```diff
@@ -0,0 +1,344 @@
+        device: PathLike = _DEFAULT_SETTINGS,
+        arch: Optional[List[Tuple[str, str]]] = None,
+        username: Optional[str] = None,
+        api_key: Optional[str] = None,
+    ):
+        """
+        Construct a new IQM backend.
+
+        Requires a valid username and API key. These can either be provided as
+        parameters or set in config using
+        :py:meth:`pytket.extensions.iqm.set_aqt_config`.
+
+        :param url: base URL for requests
+        :param device: path of JSON file containing device settings
+        :param arch: list of couplings between the qubits defined in the device settings
+            (default: [("QB1", "QB3"), ("QB2", "QB3"), ("QB4", "QB3"), ("QB5", "QB3")])
+        :param username: IQM username
+        :param api_key: IQM API key
+        """
+        super().__init__()
+        self._url = url
+        config = IQMConfig.from_default_config_file()
+
+        if username is None:
+            username = config.username
+        if username is None:
+            raise IqmAuthenticationError()
+        if api_key is None:
+            api_key = config.api_key
+        if api_key is None:
+            raise IqmAuthenticationError()
+
+        with open(device) as f:
+            settings = json.load(f)
+        self._client = IQMClient(
+            self._url, settings=settings, username=username, api_key=api_key
+        )
+        self._qubits = [
+            _as_node(qb) for qb in settings["subtrees"].keys() if qb.startswith("QB")
+        ]
+        self._n_qubits = len(self._qubits)
+        if arch is None:
+            arch = _DEFAULT_COUPLING
+        coupling = [(_as_node(a), _as_node(b)) for (a, b) in arch]
+        if any(qb not in self._qubits for couple in coupling for qb in couple):
+            raise ValueError("Architecture contains qubits not in device")
+        self._arch = Architecture(coupling)
+        self._backendinfo = BackendInfo(
+            type(self).__name__,
+            settings["name"],
+            __extension_version__,
+            self._arch,
+            _GATE_SET,
+        )
+
+    @property
+    def backend_info(self) -> BackendInfo:
+        return self._backendinfo
+
+    @property
+    def required_predicates(self) -> List[Predicate]:
+        return [
+            NoClassicalControlPredicate(),
+            NoFastFeedforwardPredicate(),
+            NoMidMeasurePredicate(),
+            NoSymbolsPredicate(),
+            GateSetPredicate(_GATE_SET),
+            ConnectivityPredicate(self._arch),
+        ]
+
+    def default_compilation_pass(self, optimisation_level: int = 1) -> BasePass:
+        assert optimisation_level in range(3)
+        passes = [DecomposeBoxes(), FlattenRegisters()]
+        if optimisation_level == 1:
+            passes.append(SynthesiseTket())
+        elif optimisation_level == 2:
+            passes.append(FullPeepholeOptimise())
+        passes.append(_iqm_rebase())  # to satisfy MaxTwoQubitGatesPredicate
```

Comment:
I think this only needs to be done for optimisation_level 0?

Reply:
Possibly, I will check.